pkg/cmd/roachprod/grafana/annotations.go

```diff
@@ -0,0 +1,126 @@
+// every request to authenticate with grafana.
+func newGrafanaClient(
+	ctx context.Context, host string, secure bool,
+) (*grafclient.GrafanaHTTPAPI, error) {
+	const grafanaAudience = "1063333028845-p47csl1ukrgnpnnjc7lrtrto6uqs9t37.apps.googleusercontent.com"
+	headers := map[string]string{}
+	scheme := "http"
+
+	if secure {
+		scheme = "https"
+
+		grafanaKey := os.Getenv(ServiceAccountJson)
+		if grafanaKey == "" {
+			return nil, errors.Newf("%s env variable was not found", ServiceAccountJson)
+		}
+		ts, err := idtoken.NewTokenSource(ctx, grafanaAudience, idtoken.WithCredentialsJSON([]byte(grafanaKey)))
+		if err != nil {
+			return nil, errors.Wrap(err, "Error creating GCS oauth token source from specified credential")
+		}
+		token, err := ts.Token()
+		if err != nil {
+			return nil, errors.Wrap(err, "Error getting token")
+		}
+
+		headers["Authorization"] = fmt.Sprintf("Bearer %s", token.AccessToken)
+	}
+
+	headers[httputil.ContentTypeHeader] = httputil.JSONContentType
+	cfg := &grafclient.TransportConfig{
+		Host:        host,
+		BasePath:    "/api",
+		Schemes:     []string{scheme},
+		HTTPHeaders: headers,
+	}
+
+	return grafclient.NewHTTPClientWithConfig(strfmt.Default, cfg), nil
+}
+
+// AddGrafanaAnnotation creates an HTTP client and sends a POST request to the
+// specified host name to create an annotation. If successful, it returns
+// the result.
+func AddGrafanaAnnotation(
+	ctx context.Context, host string, secure bool, req AddAnnotationRequest,
+) (string, error) {
+	// Cleanup the host name. The Grafana OpenAPI already appends the scheme
+	// and leading slash.
+	host, _ = strings.CutPrefix(host, "https://")
+	host, _ = strings.CutPrefix(host, "http://")
+	host = strings.Trim(host, "/")
+
+	httpClient, err := newGrafanaClient(ctx, host, secure)
+	if err != nil {
+		return "", err
+	}
+
+	body := &models.PostAnnotationsCmd{
+		Text:         &req.Text,
+		Tags:         req.Tags,
+		DashboardUID: req.DashboardUID,
+		Time:         req.StartTime,
+		TimeEnd:      req.EndTime,
+	}
+
+	res, err := httpClient.Annotations.PostAnnotation(body)
+	if err != nil {
+		emptyResponseBodyErr := fmt.Errorf("&{ <nil> } (*models.ErrorResponseBody) is not supported by the TextConsumer, can be resolved by supporting TextUnmarshaler interface")
+		// The Grafana OpenAPI does not correctly handle empty responses, i.e. renegotiating
+		// the content-type dynamically if there is an error with no body. This leads to a
+		// cryptic `(*models.ErrorResponseBody) is not supported by the TextConsumer` error,
+		// but is most likely a 401 authentication issue so give the caller a hint.
+		if errors.Is(err, emptyResponseBodyErr) {
+			return "", errors.Wrap(err, "AddGrafanaAnnotation failed - HINT: if issue is not transient, check that you are authenticating correctly")
+		}
+		return "", err
+	}
+	if !res.IsSuccess() {
+		return "", errors.Newf("AddGrafanaAnnotation failed with: %s", res)
+	}
+
+	return res.String(), nil
```

Comment:
How is the caller going to use the result? The stringified version of the response payload is kinda custom; useful for debugging, otherwise unclear.

Reply:
I originally added it for debugging, but you're right it's not very useful. We have a check here to see if the result was successful, so result should always be a `200` status if `err=nil`, and the returned `err` should show the result if not nil. Removed.